### PR TITLE
Avoid adding an HttpClient.Timeout message to ConnectTimeout exceptions

### LINKED ISSUE #67505: Confusing message on Http timeout
From some test failures I noticed:
```
NuGet.Protocol.Core.Types.FatalProtocolException : An error occurred while retrieving package metadata for 'Microsoft.NETFramework.ReferenceAssemblies.net48.1.0.2' from source 'dotnet-public'.
---- System.Threading.Tasks.TaskCanceledException : The request was canceled due to the configured HttpClient.Timeout of -0.001 seconds elapsing.
-------- System.TimeoutException : The operation was canceled.
------------ System.Threading.Tasks.TaskCanceledException : The operation was canceled.
---------------- System.TimeoutException : A connection could not be established within the configured ConnectTimeout.
```

Apparently the timeout was set to Timeout.InfiniteTimeSpan, which is -1 ms.  Presumably this couldn't have expired, so the message is confusing. I don't have a repro, so feel free to close if it's not important.

### PR DESCRIPTION
**Pull request by MihaZupan**
Fixes #67505

### CODE REVIEW

**Review by CarnaViire — APPROVED** [review]
LGTM, thanks!

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientTest.cs [hunk 1]
```diff
@@ -798,6 +798,48 @@ public async Task Timeout_SetTo30AndGetResponseQuickly_Success()
...
+
+            TimeoutException connectTimeoutException = Assert.IsType<TimeoutException>(e.InnerException);
+            Assert.Contains("ConnectTimeout", connectTimeoutException.Message);
+
+            Assert.Null(connectTimeoutException.InnerException);
+            Assert.DoesNotContain("42", e.ToString());
+        }
+
+        [Fact]
+        public async Task UnknownOperationCanceledException_NotWrappedInATimeoutException()
```
Comment (MihaZupan) [inline src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientTest.cs:828]: Looking into it a bit more, there was a broader issue here where we would classify any unknown OCE (not just `ConnectTimeout`) as `HttpClient.Timeout`.
I tweaked the checks somewhat to catch such cases as well.

File: src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs [hunk 2]
```diff
@@ -604,11 +604,18 @@ private void HandleFailure(Exception e, bool telemetryStarted, HttpResponseMessa
                         e = toThrow = new TaskCanceledException(oce.Message, oce.InnerException, cancellationToken);
                     }
                 }
-                else if (!pendingRequestsCts.IsCancellationRequested)
+                else if (cts.IsCancellationRequested && !pendingRequestsCts.IsCancellationRequested)
                 {
                     // If this exception is for cancellation, but cancellation wasn't requested, either by the caller's token or by the pending requests source,
```
Comment (CarnaViire) [inline src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs:609]:
```suggestion
                    // If the linked cancellation token source was canceled, but cancellation wasn't requested, either by the caller's token or by the pending requests source,
```

**Review by CarnaViire — APPROVED** [review]
Thanks!

**Comment by MihaZupan** [comment]
Failures are #72428

**Pull request merged**